Use file processor for user cover photos

Cyperghost opened the pull request:
Required https://github.com/WoltLab/WCF/pull/6047
See https://github.com/WoltLab/WCF/issues/5723

### TODO

- [x] Change the base branch to `6.2` after https://github.com/WoltLab/WCF/pull/6047 has been merged
- [ ] Update docs for deprecated and removed function/classes

Cyperghost committed:
Add cover photo management

Cyperghost committed:
Only GET for this form is required

Cyperghost committed:
Remove cover photo file strategy classes

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

Cyperghost committed:
Use `FileRuntimeCache`

Cyperghost committed:
Manage cover photo in acp

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

Cyperghost committed:
Dont use `coverPhotoHash` or `coverPhotoExtension` anymore

Cyperghost committed:
Insert back `IUserCoverPhoto::delete()`

Cyperghost committed:
Migrate old cover photos into the new file system

Cyperghost committed:
Check if `coverPhotoHash` or `coverPhotoExtension` is set

Cyperghost committed:
Add phpdoc to `getCoverPhotoLocation()`

Cyperghost committed:
Delete old user cover photo upload/delete

Cyperghost committed:
Add foreign key for `wcf1_user.coverPhotoFileID` to `wcf1_file.fileID`

Cyperghost committed:
Mark `IWebpUserCoverPhoto` as deprecated

Cyperghost committed:
Cache cover photos

dtdesign commented on ts/WoltLabSuite/Core/Bootstrap.ts:
```diff
@@ -52,6 +52,7 @@ window.__wcf_bc_eventHandler = EventHandler;
 export interface BoostrapOptions {
   dynamicColorScheme: boolean;
   enableMobileMenu: boolean;
+  defaultCoverPhoto?: string;
```
This doesn’t belong here, instead it should be provided on the button to edit the cover photo instead.

dtdesign commented on ts/WoltLabSuite/Core/Component/User/CoverPhoto.ts:
```diff
@@ -0,0 +1,89 @@
+      FormBuilderManager.unregisterForm(json.formId);
+    }
+
+    if (oldCoverPhoto === coverPhotoUrl || oldCoverPhoto === `url("${coverPhotoUrl}")`) {
```
The URL needs to be encoded. Ideally the entire string is stored in a variable because you are constructing it two more times below.

dtdesign commented on wcfsetup/install/files/lib/system/user/command/SetCoverPhoto.class.php:
```diff
@@ -0,0 +1,76 @@
+            (new FileAction([$this->user->coverPhotoFileID], 'delete'))->executeAction();
+        }
+
+        // Delete the old cover photo if it exists.
```
Why? The existing cover photo should be migrated as part of the `UserRebuildDataWorker`, therefore this code is either harmful or does nothing?

dtdesign commented on wcfsetup/install/files/lib/system/worker/UserRebuildDataWorker.class.php:
```diff
@@ -289,48 +288,19 @@ public function execute()
-                }
+            foreach ($userProfiles as $user) {
+                $file = FileEditor::createFromExistingFile(
+                    SetCoverPhoto::getCoverPhotoLocation($user, false),
```
This is an odd way of doing it, commands should not have publicly callable methods. If you need a helper method for the legacy location then add those to the `UserCoverPhoto` class as `getLegacyLocation(…)`.

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

Cyperghost committed:
`createMinMax` renamed to `forMinMax`

Cyperghost committed:
Rename `getCoverPhotoLocation()` to `getLegacyLocation()` and move to `UserCoverPhoto`.

Cyperghost committed:
Remove `defaultCoverPhoto` from bootstrap

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

Cyperghost committed:
`coverPhotoStyle` need to be encoded

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

# Conflicts:
#	com.woltlab.wcf/templates/user.tpl
#	ts/WoltLabSuite/Core/Bootstrap.ts
#	ts/WoltLabSuite/Core/Ui/User/CoverPhoto/Delete.ts
#	ts/WoltLabSuite/Core/Ui/User/CoverPhoto/Upload.ts
#	wcfsetup/install/files/js/WoltLabSuite/Core/Bootstrap.js
#	wcfsetup/install/files/js/WoltLabSuite/Core/Ui/User/CoverPhoto/Delete.js
#	wcfsetup/install/files/js/WoltLabSuite/Core/Ui/User/CoverPhoto/Upload.js

Cyperghost committed:
Use the user profile header template

Cyperghost committed:
Use `File::getFullSizeImageSource()` instead of `File::getLink()`

Cyperghost committed:
Add `simpleReplace` and `hideDeleteButton` to the `FileProcessorFormField`

Cyperghost committed:
Display thumbnail size by file processor

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

Cyperghost committed:
Unset `uploadResolve`

Cyperghost committed:
Merge branch 'file-upload-image-crop' into 6.2-user-coverphoto

Cyperghost committed:
Use `simpleReplace()`

Cyperghost committed:
Add a callback that is triggered when the value of a FileProcessorFormField is changed.

Cyperghost committed:
Notify only when the value has really changed.

Cyperghost committed:
Fix crop cancel event handling

Cyperghost committed:
Change cover photo instantly when a new file is uploaded or deleted

Cyperghost committed:
Add thumbnail

Cyperghost committed:
Add `IUserCoverPhoto::getThumbnailURL()`

Cyperghost committed:
Use `IUserCoverPhoto::getThumbnailURL()` in `userCard` template

Cyperghost committed:
Set small thumbnail size to 800x200

dtdesign committed:
Clamp the values when the selection violates the boundaries

dtdesign committed:
Reverse the event logic

This also avoids calling into `$moveTo` from within the `change` event which could cause a nasty loop.

Cyperghost committed:
use `$change` instead of changing all values individually, for each value the `$change` function was called

Cyperghost committed:
Do not use `precise` anymore so that we only calculate with integers

Cyperghost committed:
Merge branch '6.2' into 6.2-user-coverphoto

# Conflicts:
#	com.woltlab.wcf/fileDelete.xml
#	com.woltlab.wcf/objectType.xml
#	wcfsetup/install/files/acp/database/update_com.woltlab.wcf_6.2.php
#	wcfsetup/install/files/lib/data/user/UserProfile.class.php
#	wcfsetup/install/files/lib/data/user/UserProfileAction.class.php
#	wcfsetup/install/files/lib/data/user/UserProfileList.class.php
#	wcfsetup/setup/db/install.sql

Cyperghost committed:
Run `tsc`
